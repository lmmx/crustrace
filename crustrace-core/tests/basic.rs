use insta::assert_snapshot;
use quote::quote;
use proc_macro2::TokenStream;
use crustrace_core::trace_all_impl;
use rust_format::{Formatter, RustFmt};

fn apply_trace_all(input: TokenStream) -> String {
    let output = trace_all_impl(input);
    RustFmt::default()
        .format_tokens(output)
        .unwrap_or_else(|e| panic!("Format error: {}", e))
}

#[test]
fn test_single_function() {
    let input = quote! {
        fn hello() {
            println!("world");
        }
    };
    
    assert_snapshot!(apply_trace_all(input));
}

#[test] 
fn test_multiple_functions() {
    let input = quote! {
        fn foo(x: i32) -> i32 {
            bar(x + 1)
        }
        
        fn bar(y: i32) -> i32 {
            y * 2
        }
    };
    
    assert_snapshot!(apply_trace_all(input));
}

#[test]
fn test_generic_function() {
    let input = quote! {
        fn generic<T: Clone>(value: T) -> T {
            value.clone()
        }
    };
    
    assert_snapshot!(apply_trace_all(input));
}

#[test]
fn test_ignores_non_functions() {
    let input = quote! {
        const x: String = "fn not_a_function";
        struct Foo { field: i32 }
        fn actual_function() {}
    };
    
    assert_snapshot!(apply_trace_all(input));
}
