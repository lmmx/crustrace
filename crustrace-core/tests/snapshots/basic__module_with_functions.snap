---
source: crustrace-core/tests/basic.rs
expression: apply_trace_all(input)
---
mod calculations {
    pub fn fibonacci(n: u64) -> u64 {
        let __tracing_attr_span = tracing::span!(tracing::Level::INFO, "fibonacci", n = n);
        let __tracing_attr_guard = __tracing_attr_span.enter();
        {
            if n <= 1 {
                n
            } else {
                add_numbers(fibonacci(n - 1), fibonacci(n - 2))
            }
        }
    }
    fn add_numbers(a: u64, b: u64) -> u64 {
        let __tracing_attr_span = tracing::span!(tracing::Level::INFO, "add_numbers", a = a, b = b);
        let __tracing_attr_guard = __tracing_attr_span.enter();
        {
            a + b
        }
    }
}
